### PR TITLE
bugfix: copy data before put it into ringbuf

### PR DESCRIPTION
**Pull request by fuweid**
if the consumer pops the data slower than the producer, the data will be
overrided by the coming data.

Signed-off-by: Wei Fu <fhfuwei@163.com>

### Ⅰ. Describe what this PR did

Fixed bug caused by the shared data.

In the [pouch](https://github.com/alibaba/pouch/blob/master/daemon/containerio/cio.go#L125), we use the following statement to copy data from fifo into containerio.

```
// https://github.com/alibaba/pouch/blob/master/daemon/containerio/cio.go#L125
go func(r io.ReadCloser) {
		io.Copy(ioset.out, r)
		r.Close()
		wg.Done()
}(f)

// https://github.com/alibaba/pouch/blob/master/daemon/containerio/container_io.go#L219
func (cio *ContainerIO) Write(data []byte) (int, error) {
	switch cio.typ {
	case stdout:
		for _, b := range cio.backends {
			if cover := b.outRing.Push(data); cover {
				logrus.Warnf("cover data, backend: %s, id: %s", b.backend.Name(), cio.id)
			}
		}
	case stderr:
		for _, b := range cio.backends {
			if cover := b.errRing.Push(data); cover {
				logrus.Warnf("cover data, backend: %s, id: %s", b.backend.Name(), cio.id)
			}
		}
	}

	return len(data), nil
}
```

However, the data is shared by the multiple write. Please check the [io.Copy](https://github.com/golang/go/blob/go1.9/src/io/io.go#L378-L414).

If we don't copy the data before write it into ringbuf, it maybe override by the coming new data/

### Ⅱ. Does this pull request fix one issue?

NONE.

### Ⅲ. Describe how you did it

Please check the code.

### CODE REVIEW

**Comment by allencloud** [comment]
`daemon/containerio/container_io.go:233:55: "comming" is a misspelling of "coming"`

Big brother is watching on you. 😄 @fuweid

**Comment by fuweid** [comment]
@allencloud That is why I like the CI 😄

**Comment by allencloud** [comment]
I think the CI failure is a flaky test which is described in #1470 : 

```
FAIL: /go/src/github.com/alibaba/pouch/test/cli_restart_test.go:34: PouchRestartSuite.TestPouchRestart
/go/src/github.com/alibaba/pouch/test/cli_restart_test.go:42:
    res.Assert(c, icmd.Success)
/go/src/github.com/alibaba/pouch/vendor/github.com/gotestyourself/gotestyourself/icmd/command.go:61:
    t.Fatalf("at %s:%d - %s\n", filepath.Base(file), line, err.Error())
... Error: at cli_restart_test.go:42 - 
Command:  /usr/local/bin/pouch restart -t 1 TestPouchRestart
ExitCode: 1
Error:    exit status 1
Stdout:   
Stderr:   Error: {"message":"service endpoint with name b4f54fcf already exists"}
Failures:
ExitCode was 1 expected 0
Expected no error
```
I re-trigger the test and I will fix the test asap.

File: daemon/containerio/container_io.go [hunk 1]
```diff
@@ -228,16 +228,28 @@ func (cio *ContainerIO) Write(data []byte) (int, error) {
...
 
+	// FIXME(fuwei): In case that the data slice is reused by the writer,
+	// we should copy the data before we push it into the ringbuffer.
+	// The previous data shares the same address with the coming data.
+	// If we don't copy the data and the previous data isn't consumed by
+	// ringbuf pop action, the incoming data will override the previous data
+	// in the ringbuf.
+	//
+	// However, copy data maybe impact the performance. We need to reconsider
+	// other better way to handle the IO.
```
Comment (HusterWan) [inline daemon/containerio/container_io.go]: Yes, AWY,  we need a better way to handle IO later !

**Comment by HusterWan** [comment]
LGTM

**Pull request merged**